### DISCUSSION TITLE
Good system for "stacking" diacritics and Nth level compositions?

### DISCUSSION THREAD

**Discussion by rsms** [rsms/inter#555] [General]
As Inter v4 is getting closer to release, I'm revising the mark positioning design system for "complex" diacritics; visual compositions of two or more diacritics, like for example /dieresismacroncomb, a dieresis with a macron on top.

<img width="868" alt="Screen Shot 2023-04-09 at 11 20 13" src="https://user-images.githubusercontent.com/47046/230789833-a51032bc-55d9-47c4-9028-a2e654c7db22.png">

I'm thinking that there are two paths forward:

1. A simpler, "regular" system that essentially works like the Mark-to-mark (mkmk) feature: a glyph like /dieresismacroncomb is just a composition of /dieresis & /macron, aligned to their respective `_top` and `top` anchors. However this is not visually ideal in all situations.
2. A more complex system with custom /dieresismacroncomb.uc & /dieresismacroncomb.lc for upper-case and lower-case uses (where one of them is used for the diacritic's "combining" Unicode codepoint.) However this is quite a lot of extra work and causes some visual irregularities (ie. spacing between glyph-accent1-accent2 is different in some cases for the upper-case and lower-case versions of the same diacritic.) 

Here's a brief example of the two approaches: (click for larger image)

![mkmk-or-custom srgb](https://user-images.githubusercontent.com/47046/230790489-c785a1a3-b397-41d7-80f0-4fc434713bee.png)

Here's a more in-depth illustration of the second "custom" approach, which has two additional vertical constraints: optical center of upper-case marks and optical center of lower-case marks.

![mark-system-1 srgb](https://user-images.githubusercontent.com/47046/230790184-5a172bc2-7184-41d2-a908-b0ab25252f38.png)

What is the better path forward here?

I always tend to lean toward systematic simplicity, so my mind prefers option 1. But my heart skips a beat when I see the unbalanced spacing ("Too much air" in example above.)

**Comment by moyogo** [comment]
You can use a single /dieresiscomb_macroncomb build from /dieresiscomb and /macroncomb or have two with the additional /dieresiscomb_macroncomb.uc with positions adjusted for uppercase. It really depends on what compromise you’re happy with.

But if you want to support more cases and you’re not happy with stacking the default diacritics in general or on uppercase, I’d recommend having stacking variants.
Say you also want to support use cases like Ā̈ (/amacron/dieresiscomb), instead of adding /macroncomb_dieresiscomb and a potential /macroncomb_dieresiscomb.uc, you could have /macroncomb.stack and /dieresiscomb.stack. You can also differentiate the uppercase and lowercase variants if you want.
You need substitutions in ccmp like `sub @CombiningTopMarks @CombiningTopMarks' by @CombningTopMarks.stack; sub @CombiningTopMarks.stack @CombiningTopMarks' by @CombningTopMarks.stack;` if you use stacking variants on all glyphs or substitutions like `sub @Uppercase @CombiningTopMarks' by @CombiningTopMarks.stack; sub @CombiningTopMarks.stack @CombiningTopMarks' by @CombningTopMarks.stack;` if you only use them on uppercase for example.
How many variants you want really depends on what your happy to compromise with.

If you use https://github.com/mekkablue/ShowMarkPreview, put the anchors you want to stack and a base glyph, it will preview the stacked anchors on the base glyph, for example `/macroncomb/dieresiscomb/a/e` on one line with show the marks on the /a and /e, and `/macroncomb.uc/dieresiscomb.uc/A/E` on one line to show the marks on /A and /E.
